src/libraries/System.Private.Xml/src/System/Xml/XPath/Internal/XPathScanner.cs

```diff
@@ -39,16 +41,18 @@ public XPathScanner(string xpathExpr)
         private bool NextChar()
         {
             Debug.Assert(0 <= _xpathExprIndex && _xpathExprIndex <= _xpathExpr.Length);
-            if (_xpathExprIndex < _xpathExpr.Length)
+
+            string expr = _xpathExpr;
+            int index = _xpathExprIndex;
+            if ((uint)index < (uint)expr.Length)
```

Comment:
Just curious what is the benefit of conversion before comparison, also the use of local variables

Reply:
The main impact here is eliminating the bounds check on indexing into the string.  By performing both the comparison and indexing using locals, the JIT can trust that the values won't change between the check and the indexing, and the use of the uint casts ensures that the index is within the string (if it were negative, it would wrap around and be larger than the length).

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGueAHKAEsAbtgwwGAZQzCAdgHMGAfQT9xACwCiaqAG4+3QaPGShcjCrWadggJJyAJjAQHqPBsbETWG7FBUwAFcoWAsWP31DXncBYW9JYAgIABsGADkXDAj/ZQB5FMcACgBKaM5Yj24hADMGItV1DG1dB2cEBgAeKyaWwQA6ABkYRWayyqqKqunAkLDsyIYAXh6bXQBtRrX7JxcAaj2AXTcZquIAdgZZIJgTmYBfaI8YFNwYJ54p055lYNCRhb+ZYMADkAB0AAwgu7fJiXGrYV63D7cR4TNEeaJeUwMJKpDJZHJQZSZADupXKKKY5AhDBcgmBW2atiiEw85ks5najOszNau1cVNq9SKQQ5JS5Li69TFFhK9KgQxGCjGVK+sN+cwBROBCvWkoQxypHiZfSgbSlKwNDD2DHIMO+FyuUBuDo8aKpmv+4UWK3BULdPCdCKRDrR9yAA=

Reply:
Interesting, thanks!